src/3rd_party_bundle_add.c

```diff
@@ -94,18 +96,134 @@ static bool parse_options(int argc, char **argv)
 	return true;
 }
 
-static enum swupd_code add_bundle(char *bundle)
+static enum swupd_code create_wrapper_script(char *filename)
+{
+	enum swupd_code ret_code = 0;
+	int fd;
+	FILE *fp = NULL;
+	char *bin_directory = NULL;
+	char *script = NULL;
+	char *binary = NULL;
+	char *env_path = NULL;
+	char *env_ld_path = NULL;
+	char *third_party_bin_path = NULL;
+	char *third_party_ld_path = NULL;
+	mode_t mode = 0755;
+
+	bin_directory = third_party_get_bin_dir();
+	script = third_party_get_binary_path(sys_basename(filename));
+	binary = sys_path_join(globals.path_prefix, filename);
+
+	if (!sys_filelink_is_executable(binary)) {
+		goto close_and_exit;
+	}
+
+	if (sys_file_exists(script)) {
+		/* the binary already exists, this condition should never happen
+		 * since we checked before installing */
+		ret_code = SWUPD_UNEXPECTED_CONDITION;
+		error("There is already a binary called %s in %s, it will be skipped\n", sys_basename(filename), bin_directory);
+		goto close_and_exit;
+	}
+
+	/* if the SWUPD_3RD_PARTY_BIN_DIR does not exist, attempt to create it */
+	if (!sys_file_exists(bin_directory)) {
+		if (mkdir_p(bin_directory)) {
+			ret_code = SWUPD_COULDNT_CREATE_DIR;
+			goto close_and_exit;
+		}
+	}
+
+	if (!is_dir(bin_directory)) {
+		error("The path %s for 3rd-party content exists but is not a directory\n", bin_directory);
+		ret_code = SWUPD_COULDNT_CREATE_DIR;
+		goto close_and_exit;
+	}
+
+	/* open the file with mode set to 0755 */
+	fd = open(script, O_RDWR | O_CREAT, mode);
+	if (fd < 0) {
+		error("The file %s failed to be created\n", script);
+		ret_code = SWUPD_COULDNT_CREATE_FILE;
+		goto close_and_exit;
+	}
+	fp = fdopen(fd, "w");
+	if (!fp) {
+		error("The file %s failed to be created\n", script);
+		ret_code = SWUPD_COULDNT_CREATE_FILE;
+		goto close_and_exit;
+	}
+
+	/* get the current values of the environment */
+	env_path = getenv("PATH");
+	env_ld_path = getenv("LD_LIBRARY_PATH");
+
+	/* get the path for the 3rd-party content */
+	third_party_bin_path = str_or_die("%sbin:%susr/bin:%susr/local/bin", globals.path_prefix, globals.path_prefix, globals.path_prefix);
+	third_party_ld_path = str_or_die("%susr/lib64:%susr/local/lib64", globals.path_prefix, globals.path_prefix);
+
+	fprintf(fp, "#!/bin/bash\n");
+	fprintf(fp, "\n");
+	fprintf(fp, "export PATH=%s%s%s\n", third_party_bin_path, env_path ? ":" : "", env_path ? "$PATH" : "");
+	fprintf(fp, "export LD_LIBRARY_PATH=%s%s%s\n", third_party_ld_path, env_ld_path ? ":" : "", env_ld_path ? "$LD_LIBRARY_PATH" : "");
+	fprintf(fp, "\n");
+	fprintf(fp, "%s \"$@\"\n", binary);
```

Comment:
This is a bit hard to read. I would change that to:
```
#define WRAPPER_TEMPLATE "#!/bin/bash\n" \
                         "export PATH=%s..."

fprintf(fp, TEMPLATE, parameters...);

Reply:
done